File: Core/peer_info.cpp
#include "peer_info.h"



#include "interop.h"
#include "bitfield.h"
#include "sha1_hash.h"

using namespace Tsunami::Core;
using PeerId = Sha1Hash;

#define PEER_INT_PROP(name) \
    int PeerInfo::name::get() \
    { \
        return info_->name; \
    }

PeerInfo::PeerInfo(libtorrent::peer_info& info)
{
    info_ = new libtorrent::peer_info(info);
}

PeerInfo::~PeerInfo()
{
    delete info_;
}

unsigned int PeerInfo::flags::get()
{
    return info_->flags;
}

PEER_INT_PROP(source);
PEER_INT_PROP(up_speed);
PEER_INT_PROP(down_speed);
PEER_INT_PROP(payload_up_speed);
PEER_INT_PROP(payload_down_speed);

long long PeerInfo::total_download::get()
{
    return info_->total_download;
}

long long PeerInfo::total_upload::get()
{
    return info_->total_upload;
}

BitField ^ PeerInfo::pieces::get()
{
	return gcnew BitField(info_->pieces);
}

PeerId ^ PeerInfo::pid::get()
{
	return gcnew PeerId(info_->pid);
}

PEER_INT_PROP(upload_limit);
PEER_INT_PROP(download_limit);

System::TimeSpan PeerInfo::last_request::get()
{
    return System::TimeSpan(0, 0, (int)libtorrent::total_seconds(info_->last_request));
}

System::TimeSpan PeerInfo::last_active::get()
{
    return System::TimeSpan(0, 0, (int)libtorrent::total_seconds(info_->last_active));
}

System::TimeSpan PeerInfo::download_queue_time::get()
{
    return System::TimeSpan(0, 0, (int)libtorrent::total_seconds(info_->download_queue_time));
}

PEER_INT_PROP(queue_bytes);
PEER_INT_PROP(request_timeout);
PEER_INT_PROP(send_buffer_size);
PEER_INT_PROP(used_send_buffer);
PEER_INT_PROP(receive_buffer_size);
PEER_INT_PROP(used_receive_buffer);
PEER_INT_PROP(num_hashfails);

System::String^ PeerInfo::country::get()
{
    return interop::from_std_string(std::string(info_->country, 2));
}

System::String^ PeerInfo::inet_as_name::get()
{
	return interop::from_std_string(info_->client); //inet_as_name
}

PEER_INT_PROP(download_queue_length);
PEER_INT_PROP(timed_out_requests);
PEER_INT_PROP(busy_requests);
PEER_INT_PROP(requests_in_buffer);
PEER_INT_PROP(target_dl_queue_length);
PEER_INT_PROP(upload_queue_length);
PEER_INT_PROP(failcount);
PEER_INT_PROP(downloading_piece_index);
PEER_INT_PROP(downloading_block_index);
PEER_INT_PROP(downloading_progress);
PEER_INT_PROP(downloading_total);

System::String^ PeerInfo::client::get()
{
    return interop::from_std_string(info_->client);
}

PEER_INT_PROP(connection_type);
PEER_INT_PROP(remote_dl_rate);
PEER_INT_PROP(pending_disk_bytes);
PEER_INT_PROP(send_quota);
PEER_INT_PROP(receive_quota);
PEER_INT_PROP(rtt);
PEER_INT_PROP(num_pieces);
PEER_INT_PROP(download_rate_peak);
PEER_INT_PROP(upload_rate_peak);
PEER_INT_PROP(progress_ppm);
PEER_INT_PROP(estimated_reciprocation_rate);

File: SessionManager/Models/PeerInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Tsunami.Models
{
    public class PeerInfo
    {
        public int BusyRequests { get; set; }
        public string Client { get; set; }
        public int ConnectionType { get; set; }
        public string Country { get; set; }
        public int DownloadingBlockIndex { get; set; }
        public int DownloadingPieceIndex { get; set; }
        public int DownloadingProgress { get; set; }
        public int DownloadingTotal { get; set; }
        public int DownloadLimit { get; set; }
        public int DownloadQueueLength { get; set; }
        public TimeSpan DownloadQueueTime { get; set; }
        public int DownloadRatePeak { get; set; }
        public int DownSpeed { get; set; }
        public int EstimatedReciprocationRate { get; set; }
        public int Failcount { get; set; }
        public uint Flags { get; set; }
        public string InetAsName { get; set; }
        public TimeSpan LastActive { get; set; }
        public TimeSpan LastRequest { get; set; }
        public int NumHashfails { get; set; }
        public int NumPieces { get; set; }
        public int PayloadDownSpeed { get; set; }
        public int PayloadUpSpeed { get; set; }
        public int PendingDiskbytes { get; set; }
        public int ProgressPpm { get; set; }
        public int QueueBytes { get; set; }
        public int ReceiveBufferSize { get; set; }
        public int ReceiveQuota { get; set; }
        public int RemoteDlRate { get; set; }
        public int RequestsInBuffer { get; set; }
        public int RequestTimeout { get; set; }
        public int Rtt { get; set; }
        public int SendBufferSize { get; set; }
        public int SendQuota { get; set; }
        public int Source { get; set; }
        public int TargetDlQueueLength { get; set; }
        public int TimedOutRequests { get; set; }
        public long TotalDownload { get; set; }
        public long TotalUpload { get; set; }
        public int UploadLimit { get; set; }
        public int UploadQueueLength { get; set; }
        public int UploadRatePeak { get; set; }
        public int UpSpeed { get; set; }
        public int UsedReceiveBuffer { get; set; }
        public int UsedSendBuffer { get; set; }

        public PeerInfo() { /* nothing to do. just for serializator */ }

        public PeerInfo(Core.PeerInfo pi)
        {
            BusyRequests = pi.busy_requests;
            Client = pi.client;
            ConnectionType = pi.connection_type;
            Country = pi.country;
            DownloadingBlockIndex = pi.downloading_block_index;
            DownloadingPieceIndex = pi.downloading_piece_index;
            DownloadingProgress = pi.downloading_progress;
            DownloadingTotal = pi.downloading_total;
            DownloadLimit = pi.download_limit;
            DownloadQueueLength = pi.download_queue_length;
            DownloadQueueTime = pi.download_queue_time;
            DownloadRatePeak = pi.download_rate_peak;
            DownSpeed = pi.down_speed;
            EstimatedReciprocationRate = pi.estimated_reciprocation_rate;
            Failcount = pi.failcount;
            Flags = pi.flags;
            InetAsName = pi.inet_as_name;
            LastActive = pi.last_active;
            LastRequest = pi.last_request;
            NumHashfails = pi.num_hashfails;
            NumPieces = pi.num_pieces;
            PayloadDownSpeed = pi.payload_down_speed;
            PayloadUpSpeed = pi.payload_up_speed;
            PendingDiskbytes = pi.pending_disk_bytes;
            ProgressPpm = pi.progress_ppm;
            QueueBytes = pi.queue_bytes;
            ReceiveBufferSize = pi.receive_buffer_size;
            ReceiveQuota = pi.receive_quota;
            RemoteDlRate = pi.remote_dl_rate;
            RequestsInBuffer = pi.requests_in_buffer;
            RequestTimeout = pi.request_timeout;
            Rtt = pi.rtt;
            SendBufferSize = pi.send_buffer_size;
            SendQuota = pi.send_quota;
            Source = pi.source;
            TargetDlQueueLength = pi.target_dl_queue_length;
            TimedOutRequests = pi.timed_out_requests;
            TotalDownload = pi.total_download;
            TotalUpload = pi.total_upload;
            UploadLimit = pi.upload_limit;
            UploadQueueLength = pi.upload_queue_length;
            UploadRatePeak = pi.upload_rate_peak;
            UpSpeed = pi.up_speed;
            UsedReceiveBuffer = pi.used_receive_buffer;
            UsedSendBuffer = pi.used_send_buffer;
        }
    }
}
